psi4/src/psi4/libmints/matrix.cc

```diff
@@ -2076,18 +2081,33 @@ void Matrix::pivoted_cholesky(double tol, std::vector<std::vector<int>> &pivot)
     }
 
     // Properly sized return matrix
-    auto L = std::make_shared<Matrix>("Cholesky decomposed matrix", nirrep_, rowspi_, nchol);
-    L->zero();
-    for (int h = 0; h < nirrep_; ++h) {
-        for (int m = 0; m < rowspi_[h]; ++m) {
-            for (int n = 0; n < std::min(m + 1, nchol[h]); ++n) {
-                // Psi4 stores matrices as row major, LAPACK as column major
-                L->set(h, m, n, get(h, n, m));
+    if (upper) {
+        auto U = std::make_shared<Matrix>("Cholesky decomposed matrix", nirrep_, rowspi_, nchol);
+        U->zero();
+        for (int h = 0; h < nirrep_; ++h) {
+            for (int m = 0; m < rowspi_[h]; ++m) {
+                for (int n = m; n < nchol[h]; ++n) {
+                    // Psi4 stores matrices as row major, LAPACK as column major
+                    U->set(h, m, n, get(h, n, m));
+                }
+            }
+        }
+        // Switch to the properly sized matrix
+        *this = *U;
+    } else {
```

Comment:
Allocating an entirely new matrix just to perform a transpose seems wasteful. Can't this be done with a value swap or something. Don't we have a function that already performs this?

Reply:
Yes, you have a point, we can use `std::swap`

Reply:
Well, the reason why we allocate a new matrix is not merely to perform a transpose, but because the number of columns of the matrix changes since the rank is reduced in pivoted Cholesky